### PR TITLE
Correct illumination and stretch images based on whole plate

### LINKED ISSUE #38: Check for signal when doing image normalisation/illumination correction
Some images contain no signal in one of the channels (for example, when none of the cells express the target protein). When stretchlim is applied, the remaining background noise gets amplified, resulting in a "haze", resulting in bad visualisation and probably bad feature extraction. Coming up with a test that would suppress noise instead of applying stretchlim in these situations would fix this.

@starcalibre can you provide one good and one bad example from the public screen?

### PR DESCRIPTION
**Pull request by jni**
Fixes #38.

Or it should, anyway, haven't tested on real data yet. =P

I recommend a `--stretchlim-output` setting of 0.001 or thereabouts, or all images will be scaled to the brightest fluff. (at least if it makes it through the sampling; but presumably one or two pixels will...)

### CODE REVIEW

File: microscopium/_util.py [hunk 1]
```diff
@@ -25,3 +27,32 @@ def groupby(key, iterable, transform=None):
...
+    this writing.)
+    """
+    if seed is None:
+        return np.random.mtrand._rand
+    if isinstance(seed, (numbers.Integral, np.integer)):
+        return np.random.RandomState(seed)
+    if isinstance(seed, np.random.RandomState):
+        return seed
+    else:
+        raise ValueError('Invalid input %s to generate random state' % seed)
```
Comment (starcalibre) [inline microscopium/_util.py]: Should exceptions raised be noted in the docstring?

File: microscopium/preprocess.py [hunk 2]
```diff
@@ -569,6 +578,94 @@ def find_background_illumination(fns, radius=51, quantile=0.05,
     return illum
 
 
+def correct_multiimage_illumination(im_fns, illum, stretch_quantile=0,
+                                    random_state=None):
+    """Divide input images pointwise by the illumination field.
+
+    However, where `correct_image_illumination` rescales each individual
+    image to span the full dynamic range of the data type, this one
```
Comment (starcalibre) [inline microscopium/preprocess.py]: Maybe a reference to this function should be added to the `correct_image_illumination` docstring?

File: microscopium/preprocess.py [hunk 3]
```diff
@@ -569,6 +578,94 @@ def find_background_illumination(fns, radius=51, quantile=0.05,
...
+        out = exposure.rescale_intensity(corrected, in_range=corr_range,
+                                         out_range=np.uint8)
+        yield out
+
+
+def _reservoir_sampled_image(ims_iter, random_state=None):
+    """Return an image where each pixel is sampled from a list of images.
+
+    The idea is to get a sample of image intensity throughout a collection
+    of images, to know what the "standard range" is for this type of image.
```
Comment (starcalibre) [inline microscopium/preprocess.py]: I'd make reference to the fact you're performing reservoir sampling here, maybe the docstring could read: Perform reservoir sampling over a collection of images to estimate the "standard range" is for this type of image.

Nice implementation btw, maybe a candidate code example for your book? :)

File: microscopium/preprocess.py [hunk 4]
```diff
@@ -569,6 +578,94 @@ def find_background_illumination(fns, radius=51, quantile=0.05,
...
+
+    Returns
+    -------
+    sampled : array, same shape as input
+        The sampled "image".
+
+    Examples
+    --------
+    >>> ims = np.arange(27).reshape((3, 3, 3))
+    >>> _reservoir_sampled_image(iter(ims), 0)
```
Comment (starcalibre) [inline microscopium/preprocess.py]: I don't think it's necessary to cast ``ims`` as an iterable here? This already happens in the function..

File: tests/test_pre.py [hunk 5]
```diff
@@ -66,5 +66,49 @@ def test_illumination_median(image_files):
...
+        io.imsave(fn, im)
+
+    def cleanup():
+        for fn in files:
+            os.remove(fn)
+    request.addfinalizer(cleanup)
+
+    illum = 0.01 * np.ones(shape, dtype=float)
+    return files, illum
+
```
Comment (starcalibre) [inline tests/test_pre.py]: Really cool!

File: microscopium/preprocess.py [hunk 6]
```diff
@@ -171,7 +173,7 @@ def maxes(fns):
     return maxes
 
 
-def stretchlim(im, bottom=0.01, top=None, mask=None):
+def stretchlim(im, bottom=0.01, top=None, mask=None, in_place=False):
```
Comment (starcalibre) [inline microscopium/preprocess.py]: Maybe the default should be 0.001 here? You recommended this, and I had success using this value with the Cellomics datasets.

**Comment by starcalibre** [comment]
Cool. Can't wait to see how this turns out!

File: microscopium/_util.py (hunk 1 above)
Comment (jni) [inline microscopium/_util.py]: Only those that might be unexpected.

File: microscopium/preprocess.py (hunk 2 above)
Comment (jni) [inline microscopium/preprocess.py]: done

File: microscopium/preprocess.py (hunk 3 above)
Comment (jni) [inline microscopium/preprocess.py]: I am rather pleased with it, yes. =D Can you submit it? I want to minimise submissions from the authors. =D It'll make a nice addition to the streaming data chapter!

I won't put implementation details in the one-liner part of the docstring, but I've added a note to the additional description.

File: microscopium/preprocess.py (hunk 4 above)
Comment (jni) [inline microscopium/preprocess.py]: Not necessary, but I wanted to show that that's what's expected, not a 3D image.

Comment (jni) [inline microscopium/preprocess.py]: Actually you're right: I'll cast *when creating ims*. That's clearer.

File: microscopium/preprocess.py (hunk 6 above)
Comment (jni) [inline microscopium/preprocess.py]: good idea, especially with the new reservoir sampling (fewer chance of fluff ruining the whole normalisation).

**Comment by jni** [comment]
@starcalibre assuming Travis is happy, I believe I've addressed all your comments! Feel free to merge!

**Pull request merged**